Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp

```diff
@@ -135,332 +145,415 @@ void SourceBufferPrivateRemote::resetParserStateInternal()
+        segment.duration = segmentInfo.duration;
 
-    client->sourceBufferPrivateDidReceiveInitializationSegment(WTFMove(segment))->whenSettled(RunLoop::current(), WTFMove(completionHandler));
+        segment.audioTracks = WTF::map(segmentInfo.audioTracks, [&](auto& audioTrack) {
+            SourceBufferPrivateClient::InitializationSegment::AudioTrackInformation info {
+                RemoteMediaDescription::create(audioTrack.description),
+                mediaPlayer->audioTrackPrivateRemote(audioTrack.id)
+            };
+            return info;
+        });
+
+        segment.videoTracks = WTF::map(segmentInfo.videoTracks, [&](auto& videoTrack) {
+            SourceBufferPrivateClient::InitializationSegment::VideoTrackInformation info {
+                RemoteMediaDescription::create(videoTrack.description),
+                mediaPlayer->videoTrackPrivateRemote(videoTrack.id)
+            };
+            return info;
+        });
+
+        segment.textTracks = WTF::map(segmentInfo.textTracks, [&](auto& textTrack) {
+            SourceBufferPrivateClient::InitializationSegment::TextTrackInformation info {
+                RemoteMediaDescription::create(textTrack.description),
+                mediaPlayer->textTrackPrivateRemote(textTrack.id)
+            };
+            return info;
+        });
+
+        return client->sourceBufferPrivateDidReceiveInitializationSegment(WTFMove(segment));
+    })->whenSettled(SourceBufferPrivateRemote::queue(), WTFMove(completionHandler));
 }
 
-void SourceBufferPrivateRemote::sourceBufferPrivateHighestPresentationTimestampChanged(const MediaTime& timestamp)
+void SourceBufferPrivateRemote::MessageReceiver::sourceBufferPrivateHighestPresentationTimestampChanged(const MediaTime& timestamp)
 {
-    if (RefPtr client = this->client())
-        client->sourceBufferPrivateHighestPresentationTimestampChanged(timestamp);
+    assertIsCurrent(SourceBufferPrivateRemote::queue());
+
+    if (RefPtr parent = m_parent.get()) {
+        if (RefPtr client = parent->client())
+            client->sourceBufferPrivateHighestPresentationTimestampChanged(timestamp);
+    }
 }
 
-void SourceBufferPrivateRemote::sourceBufferPrivateDurationChanged(const MediaTime& duration, CompletionHandler<void()>&& completionHandler)
+void SourceBufferPrivateRemote::MessageReceiver::sourceBufferPrivateDurationChanged(const MediaTime& duration, CompletionHandler<void()>&& completionHandler)
 {
-    if (RefPtr client = this->client())
-        client->sourceBufferPrivateDurationChanged(duration)->whenSettled(RunLoop::current(), WTFMove(completionHandler));
-    else
-        completionHandler();
+    assertIsCurrent(SourceBufferPrivateRemote::queue());
+
+    if (RefPtr parent = m_parent.get()) {
+        if (RefPtr client = parent->client()) {
+            client->sourceBufferPrivateDurationChanged(duration)->whenSettled(SourceBufferPrivateRemote::queue(), WTFMove(completionHandler));
+            return;
+        }
+    }
```

Comment:
Ditto. This pattern shows up enough, I wonder if it needs a helper `client()` method.